File: sdk/ts/src/resolution/reverseResolver.ts
import { PublicKey, publicKey } from '@metaplex-foundation/umi';
import { normalizeChainCaip2 } from './caip';
import { DEFAULT_SOLANA_CAIP2, SRS_DEFAULT_PROGRAM_ID } from './constants';
import { ResolutionInputError } from './errors';
import { normalizeName } from './namehash';
import { findRecordPda, reverseRecordSeed } from './pda';
import type { RecordAccountProvider } from './provider';
import {
  parseResolutionTuples,
  type ResolutionTuple,
} from './tupleCodec';

export interface ForwardNameResolver {
  resolve(name: string, chainCaip2?: string): Promise<string | null>;
}

export interface SrsReverseResolverConfig {
  provider: RecordAccountProvider;
  reverseClassAddress: PublicKey;
  programId?: PublicKey;
  defaultChainCaip2?: string;
  verifyReverseWithForward?: boolean;
  forwardVerifier?: ForwardNameResolver;
}

export class SrsReverseResolver {
  private readonly provider: RecordAccountProvider;
  private readonly reverseClassAddress: PublicKey;
  private readonly programId: PublicKey;
  private readonly defaultChainCaip2: string;
  private readonly verifyReverseWithForward: boolean;
  private readonly forwardVerifier?: ForwardNameResolver;

  constructor(config: SrsReverseResolverConfig) {
    this.provider = config.provider;
    this.reverseClassAddress = config.reverseClassAddress;
    this.programId = config.programId ?? SRS_DEFAULT_PROGRAM_ID;
    this.defaultChainCaip2 = normalizeChainCaip2(
      config.defaultChainCaip2 ?? DEFAULT_SOLANA_CAIP2
    );
    this.verifyReverseWithForward = config.verifyReverseWithForward ?? true;
    this.forwardVerifier = config.forwardVerifier;

    if (this.verifyReverseWithForward && !this.forwardVerifier) {
      throw new ResolutionInputError(
        'forwardVerifier is required when verifyReverseWithForward is enabled'
      );
    }
  }

  async reverseResolve(wallet: string): Promise<string | null> {
    const normalizedWallet = publicKey(wallet);
    const tuples = await this.fetchReverseTuples(normalizedWallet);
    if (!tuples) {
      return null;
    }

    let selectedName: string | null = null;
    for (const name of this.extractReverseNames(tuples)) {
      selectedName = name;
    }

    if (!selectedName) {
      return null;
    }

    if (!this.verifyReverseWithForward) {
      return selectedName;
    }

    const resolvedWallet = await this.forwardVerifier?.resolve(
      selectedName,
      this.defaultChainCaip2
    );
    if (!resolvedWallet) {
      return null;
    }

    return resolvedWallet === normalizedWallet ? selectedName : null;
  }

  async reverseResolveAll(wallet: string): Promise<string[]> {
    const normalizedWallet = publicKey(wallet);
    const tuples = await this.fetchReverseTuples(normalizedWallet);
    if (!tuples) {
      return [];
    }

    const names = this.extractReverseNames(tuples);
    if (!this.verifyReverseWithForward) {
      return names;
    }

    const verified = await Promise.all(
      names.map(async (name) => {
        const resolvedWallet = await this.forwardVerifier?.resolve(
          name,
          this.defaultChainCaip2
        );
        return resolvedWallet === normalizedWallet ? name : null;
      })
    );
    return verified.filter((name): name is string => name !== null);
  }

  async batchReverseResolve(
    wallets: readonly string[]
  ): Promise<Array<string | null>> {
    return Promise.all(wallets.map((wallet) => this.reverseResolve(wallet)));
  }

  async batchReverseResolveAll(
    wallets: readonly string[]
  ): Promise<string[][]> {
    return Promise.all(wallets.map((wallet) => this.reverseResolveAll(wallet)));
  }

  private extractReverseNames(tuples: readonly ResolutionTuple[]): string[] {
    const names: string[] = [];
    for (const [key, value] of tuples) {
      if (key.trim().toUpperCase() !== 'NAME') {
        continue;
      }

      const candidateName = value.trim();
      if (candidateName.length > 0) {
        names.push(normalizeName(candidateName));
      }
    }

    return names;
  }

  private async fetchReverseTuples(wallet: string): Promise<ResolutionTuple[] | null> {
    const reverseSeed = reverseRecordSeed(wallet);
    const [recordPda] = await findRecordPda(
      this.reverseClassAddress,
      reverseSeed,
      this.programId
    );

    return this.fetchRecordTuples(recordPda);
  }

  private async fetchRecordTuples(recordPda: PublicKey) {
    const record = await this.provider.fetchRecord(recordPda);
    if (!record) {
      return null;
    }

    return parseResolutionTuples(record.data);
  }
}

File: sdk/rust/src/resolution/reverse_resolver.rs
use std::sync::Arc;

use solana_program::pubkey::Pubkey;

use crate::resolution::caip::normalize_chain_caip2;
use crate::resolution::constants::{DEFAULT_SOLANA_CAIP2, SRS_DEFAULT_PROGRAM_ID};
use crate::resolution::errors::{ResolutionError, ResolutionResult};
use crate::resolution::forward_resolver::ForwardNameResolver;
use crate::resolution::namehash::normalize_name;
use crate::resolution::pda::{find_record_pda, reverse_record_seed};
use crate::resolution::provider::RawRecordAccountProvider;
use crate::resolution::record_decoder::decode_srs_record;
use crate::resolution::tuple_codec::{parse_resolution_tuples, ResolutionTuple};

pub struct SrsReverseResolverConfig {
    pub provider: Arc<dyn RawRecordAccountProvider>,
    pub reverse_class_address: Pubkey,
    pub program_id: Option<Pubkey>,
    pub default_chain_caip2: Option<String>,
    pub verify_reverse_with_forward: Option<bool>,
    pub forward_verifier: Option<Arc<dyn ForwardNameResolver>>,
}

pub struct SrsReverseResolver {
    provider: Arc<dyn RawRecordAccountProvider>,
    reverse_class_address: Pubkey,
    program_id: Pubkey,
    default_chain_caip2: String,
    verify_reverse_with_forward: bool,
    forward_verifier: Option<Arc<dyn ForwardNameResolver>>,
}

impl SrsReverseResolver {
    pub fn new(config: SrsReverseResolverConfig) -> ResolutionResult<Self> {
        let default_chain_caip2 = normalize_chain_caip2(
            config
                .default_chain_caip2
                .as_deref()
                .unwrap_or(DEFAULT_SOLANA_CAIP2),
        )?;

        let verify_reverse_with_forward = config.verify_reverse_with_forward.unwrap_or(true);
        if verify_reverse_with_forward && config.forward_verifier.is_none() {
            return Err(ResolutionError::Input(
                "forward_verifier is required when verify_reverse_with_forward is enabled"
                    .to_string(),
            ));
        }

        Ok(Self {
            provider: config.provider,
            reverse_class_address: config.reverse_class_address,
            program_id: config.program_id.unwrap_or(SRS_DEFAULT_PROGRAM_ID),
            default_chain_caip2,
            verify_reverse_with_forward,
            forward_verifier: config.forward_verifier,
        })
    }

    pub fn reverse_resolve(&self, wallet: &str) -> ResolutionResult<Option<String>> {
        let reverse_seed = reverse_record_seed(wallet)?;
        let (record_pda, _) =
            find_record_pda(&self.reverse_class_address, &reverse_seed, &self.program_id)?;

        let Some(tuples) = self.fetch_record_tuples(&record_pda)? else {
            return Ok(None);
        };

        let mut selected_name: Option<String> = None;
        for (key, value) in tuples {
            if key.trim().to_uppercase() != "NAME" {
                continue;
            }

            let candidate_name = value.trim();
            if candidate_name.is_empty() {
                continue;
            }

            selected_name = Some(normalize_name(candidate_name)?);
        }

        let Some(selected_name) = selected_name else {
            return Ok(None);
        };

        if !self.verify_reverse_with_forward {
            return Ok(Some(selected_name));
        }

        let forward_verifier = self
            .forward_verifier
            .as_ref()
            .expect("forward verifier is validated in constructor");
        let resolved_wallet =
            forward_verifier.resolve_forward(&selected_name, Some(&self.default_chain_caip2))?;

        Ok(match resolved_wallet {
            Some(wallet_from_forward) if wallet_from_forward == wallet => Some(selected_name),
            _ => None,
        })
    }

    pub fn batch_reverse_resolve<S: AsRef<str>>(
        &self,
        wallets: &[S],
    ) -> ResolutionResult<Vec<Option<String>>> {
        wallets
            .iter()
            .map(|wallet| self.reverse_resolve(wallet.as_ref()))
            .collect()
    }

    fn fetch_record_tuples(
        &self,
        record_pda: &Pubkey,
    ) -> ResolutionResult<Option<Vec<ResolutionTuple>>> {
        let Some(raw) = self.provider.fetch_raw_record_account(record_pda)? else {
            return Ok(None);
        };

        let decoded = decode_srs_record(&raw)?;
        Ok(Some(parse_resolution_tuples(&decoded.data)?))
    }
}
